File: src/Json/Process/AssayMaterials.fs
namespace ARCtrl.Json

open Thoth.Json.Core

open ARCtrl
open ARCtrl.Process
open System.IO

module AssayMaterials = 

    module ISAJson = 

        let encoder (ps : Process list) = 
            let samples = ProcessSequence.getSamples ps
            let materials = ProcessSequence.getMaterials ps
            [
                Encode.tryIncludeList "samples" Sample.ISAJson.encoder samples
                Encode.tryIncludeList "otherMaterials" Material.ISAJson.encoder materials
            ]
           
            |> Encode.choose
            |> Encode.object
    
        // No decoder, as this is a write-only operation

        //let allowedFields = ["samples";"otherMaterials"]

        //let decoder (options : ConverterOptions) : Decoder<AssayMaterials> =
        //    Decode.object allowedFields (fun get ->
        //        {
        //            Samples = get.Optional.Field "samples" (Decode.list (Sample.decoder options))
        //            OtherMaterials = get.Optional.Field "otherMaterials" (Decode.list (Material.decoder options))
        //        }
        //    )

File: js/Json/Process/AssayMaterials.js
import { getMaterials, getSamples } from "../../Core/Process/ProcessSequence.js";
import { ofArray, choose } from "../../fable_modules/fable-library-js.4.22.0/List.js";
import { map } from "../../fable_modules/fable-library-js.4.22.0/Option.js";
import { tryIncludeList } from "../Encode.js";
import { ISAJson_encoder } from "./Sample.js";
import { ISAJson_encoder as ISAJson_encoder_1 } from "./Material.js";
import { map as map_1 } from "../../fable_modules/fable-library-js.4.22.0/Seq.js";

export function encoder(idMap, ps) {
    const samples = getSamples(ps);
    const materials = getMaterials(ps);
    const values = choose((tupledArg) => map((v_1) => [tupledArg[0], v_1], tupledArg[1]), ofArray([tryIncludeList("samples", (oa) => ISAJson_encoder(idMap, oa), samples), tryIncludeList("otherMaterials", (c) => ISAJson_encoder_1(idMap, c), materials)]));
    return {
        Encode(helpers) {
            const arg = map_1((tupledArg_1) => [tupledArg_1[0], tupledArg_1[1].Encode(helpers)], values);
            return helpers.encodeObject(arg);
        },
    };
}

